openstack_hypervisor/fileserver/server.py

```diff
@@ -0,0 +1,398 @@
+        return create_dir_ep(request)
+    if request.path_info == "/fs/remove-dir" and request.method == "DELETE":
+        return remove_dir_ep(request)
+    if request.path_info == "/upload/init" and request.method == "POST":
+        return upload_init_ep(request)
+    if request.path_info.startswith("/upload/") and request.method == "PUT":
+        upload_id = request.path_info.split("/", 2)[2]
+        try:
+            chunk_index = int(request.GET.get("chunk_index", ""))
+            total_chunks = int(request.GET.get("total_chunks", ""))
+        except ValueError:
+            return _error(400, "chunk_index and total_chunks must be integers")
+        return upload_chunk_ep(request, upload_id, chunk_index, total_chunks)
+    if request.path_info == "/upload/finalize" and request.method == "POST":
+        return upload_finalize_ep(request)
+    if request.path_info == "/download" and request.method == "GET":
+        return download_file_ep(request)
+    return _error(404, "Not found")
+
+
+def application(environ, start_response):
+    """WSGI application callable."""
+    request = Request(environ)
+    response = _route(request)
+    return response(environ, start_response)
+
+
+class ThreadingWSGIServer(ThreadingMixIn, WSGIServer):
+    """Threading-based WSGI server."""
+
+    daemon_threads = True
+
+
+class ThreadingWSGIService(service.ServiceBase):
+    """Threading-based WSGI service."""
+
+    def __init__(self, app, host: str, port: int, ssl_context: ssl.SSLContext | None):
+        self._app = app
+        self._host = host
+        self._port = port
+        self._ssl_context = ssl_context
+        self._httpd = None
+        self._thread = None
+
+    def start(self):
+        """Start the WSGI service."""
+        self._httpd = make_server(
+            self._host, self._port, self._app, server_class=ThreadingWSGIServer
+        )
+        if self._ssl_context is not None:
+            self._httpd.socket = self._ssl_context.wrap_socket(
+                self._httpd.socket, server_side=True
+            )
+        self._thread = threading.Thread(
+            target=self._httpd.serve_forever, name="fileserver", daemon=True
+        )
+        self._thread.start()
+
+    def stop(self, graceful=True):
+        """Stop the WSGI service."""
+        if self._httpd is not None:
+            self._httpd.shutdown()
+
+    def wait(self):
+        """Wait for the WSGI service to finish."""
+        if self._thread is not None:
+            self._thread.join()
+
+    def reset(self, exiting=False):
+        """Reset service state (no-op)."""
+        return
+
+
+if __name__ == "__main__":
+    logging.register_options(CONF)
+    logging.setup(CONF, "openstack_hypervisor.fileserver")
+    CONF(
+        project="openstack-hypervisor",
+        prog="openstack-hypervisor-fileserver",
+        version="1.0.0",
+    )
+
+    ssl_cert_path = str(get_snap_common() / "etc/pki/nova/servercert.pem")
```

Comment:
try not to use getting snap related paths in fileserver and try to get full path as part of configuration parameters. Aim is to able to run fileserver outside snap.

Reply:
Hemanth is right, in the implementation of the file server, no mention to snap whatsoever should be made. The goal is to upstream that code later on. Let's avoid a full rewrite for something we already know.

Only the snap hooks will configure the config options setting up the right paths from a snap POV.